File: leetcode/0977_squares_of_a_sorted_array.cpp
// Copyright (c) 2019 Jacopo Notarstefano

#include <algorithm>
#include <cassert>
#include <cmath>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<int> sortedSquares(vector<int>& A) {
    vector<int> result;

    int i = 0, j = A.size() - 1;
    while (i <= j) {
      if (abs(A[i]) < abs(A[j])) {
        result.push_back(A[j] * A[j]);
        j--;
      } else {
        result.push_back(A[i] * A[i]);
        i++;
      }
    }

    reverse(result.begin(), result.end());

    return result;
  }
};

int main() {
  auto solution = Solution();

  vector<int> A = {-4, -1, 0, 3, 10};

  vector<int> expected = {0, 1, 9, 16, 100};
  vector<int> result = solution.sortedSquares(A);

  assert(expected == result);
}

File: leetcode/0977_squares_of_a_sorted_array.py
# -*- coding: utf-8 -*-


class Solution:
    def sortedSquares(self, A):
        result = []

        i, j = 0, len(A) - 1
        while i <= j:
            if abs(A[i]) < abs(A[j]):
                result.append(A[j] * A[j])
                j -= 1
            else:
                result.append(A[i] * A[i])
                i += 1

        return result[::-1]


if __name__ == '__main__':
    solution = Solution()

    assert [0, 1, 9, 16, 100] == solution.sortedSquares([-4, -1, 0, 3, 10])
    assert [4, 9, 9, 49, 121] == solution.sortedSquares([-7, -3, 2, 3, 11])
